src/gateway/sharding/shard_manager.rs

```diff
@@ -390,14 +379,70 @@ pub struct ShardManagerOptions {
     #[cfg(feature = "voice")]
     pub voice_manager: Option<Arc<dyn VoiceGatewayManager>>,
     pub ws_url: Arc<str>,
+    pub compression: TransportCompression,
     pub shard_total: NonZeroU16,
+    pub max_concurrency: NonZeroU16,
+    pub wait_time_between_shard_start: Duration,
     #[cfg(feature = "cache")]
     pub cache: Arc<Cache>,
     pub http: Arc<Http>,
     pub intents: GatewayIntents,
     pub presence: Option<PresenceData>,
-    pub max_concurrency: NonZeroU16,
-    /// Number of seconds to wait between starting each shard/set of shards start
-    pub wait_time_between_shard_start: Duration,
-    pub compression: TransportCompression,
+}
+
+pub enum ShardManagerMessage {
+    PollShardQueue,
+    Restart(ShardId),
+    Err(GatewayError),
+}
+
+/// A queue of [`ShardId`]s that is split up into multiple buckets according to the value of
+/// [`max_concurrency`](crate::model::gateway::SessionStartLimit::max_concurrency).
+#[must_use]
+pub struct ShardQueue {
+    buckets: FixedArray<VecDeque<ShardId>, u16>,
+    concurrent: bool,
+}
+
+impl ShardQueue {
+    pub fn new(max_concurrency: NonZeroU16) -> Self {
+        let buckets = vec![VecDeque::new(); max_concurrency.get() as usize].into_boxed_slice();
+        let buckets = FixedArray::try_from(buckets).expect("should fit without truncation");
+
+        Self {
+            buckets,
+            concurrent: false,
+        }
+    }
+
+    fn calculate_bucket(&self, shard_id: ShardId) -> u16 {
+        shard_id.0 % self.buckets.len()
+    }
+
+    /// Calculates the corresponding bucket for the given `ShardId` and **appends** to it.
+    pub fn push_back(&mut self, shard_id: ShardId) {
+        let bucket = self.calculate_bucket(shard_id);
+        self.buckets[bucket].push_back(shard_id);
+    }
+
+    /// Calculates the corresponding bucket for the given `ShardId` and **prepends** to it.
+    pub fn push_front(&mut self, shard_id: ShardId) {
+        let bucket = self.calculate_bucket(shard_id);
+        self.buckets[bucket].push_front(shard_id);
+    }
+
+    /// Pops a `ShardId` from every bucket containing at least one and returns them all as a `Vec`.
+    /// If the queue is in concurrent mode, will wait until all buckets are filled before returning
```

Comment:
"will wait" makes me think this method is doing some kind of async locking/waiting, when it instead just returns `None`. Can this wording be improved?